server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -471,7 +471,7 @@ public void updateShardState(final ShardRouting newRouting,
                                 final Engine engine = getEngine();
                                 engine.restoreLocalHistoryFromTranslog((resettingEngine, snapshot) ->
                                     runTranslogRecovery(resettingEngine, snapshot, Engine.Operation.Origin.LOCAL_RESET, () -> {}));
-                                if (indexSettings.getIndexVersionCreated().onOrBefore(Version.V_3_0_1)) {
+                                if (indexSettings.getIndexVersionCreated().onOrBefore(Version.CURRENT.minimumCompatibilityVersion())) {
```

Comment:
I think we could replace this with an assert that the version is >= 4_0_0 and drop the if branch